plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -408,14 +414,16 @@ private List<DeleteFilter> readDeletes(
                 }
             }
             else if (delete.content() == EQUALITY_DELETES) {
-                List<Integer> fieldIds = delete.equalityFieldIds();
+                Set<Integer> fieldIds = Sets.newHashSet(delete.equalityFieldIds());
                 verify(!fieldIds.isEmpty(), "equality field IDs are missing");
-                List<IcebergColumnHandle> columns = fieldIds.stream()
-                        .map(id -> getColumnHandle(schema.findField(id), typeManager))
-                        .collect(toImmutableList());
+                Schema deleteSchema = TypeUtil.select(schema, fieldIds);
+                List<IcebergColumnHandle> columns = deleteSchema.columns().stream()
+                                .map(column -> getColumnHandle(column, typeManager)).collect(toImmutableList());
+
+                StructLikeSet deleteSet = equalityDeleteSets.getDeleteSet(fieldIds, deleteSchema);
 
                 try (ConnectorPageSource pageSource = openDeletes(session, delete, columns, TupleDomain.all())) {
-                    filters.add(readEqualityDeletes(pageSource, columns, schema));
+                    readEqualityDeletes(pageSource, columns, deleteSet, schema);
```

Comment:
Can be done as a follow-up: The `readEqualityDeletes` would fit better within `IcebergPageSourceProvider` instead of inside `EqualityDeleteFilter`.

Reply:
PositionDeleteFilter has same problem